VultisigApp/VultisigApp/Services/BlockChainService.swift

```diff
@@ -37,136 +41,175 @@ final class BlockChainService {
             async let highPriorityFeePromise = sol.fetchHighPriorityFee(account: coin.address)
-            
+
             let recentBlockHash = try await recentBlockHashPromise
             let highPriorityFee = try await highPriorityFeePromise
-            
+
             guard let recentBlockHash else {
                 throw Errors.failToGetRecentBlockHash
             }
+
+            if let fromAddress, let toAddress, !coin.isNativeToken {
+                async let associatedTokenAddressFromPromise = sol.fetchTokenAssociatedAccountByOwner(for: fromAddress, mintAddress: coin.contractAddress)
+                async let associatedTokenAddressToPromise = sol.fetchTokenAssociatedAccountByOwner(for: toAddress, mintAddress: coin.contractAddress)
+                let associatedTokenAddressFrom = try await associatedTokenAddressFromPromise
+                let associatedTokenAddressTo = try await associatedTokenAddressToPromise
+
+                return .Solana(recentBlockHash: recentBlockHash, priorityFee: BigInt(highPriorityFee), fromAddressPubKey: associatedTokenAddressFrom, toAddressPubKey: associatedTokenAddressTo)
+            }
+
             return .Solana(recentBlockHash: recentBlockHash, priorityFee: BigInt(highPriorityFee), fromAddressPubKey: nil, toAddressPubKey: nil)
-            
+
         case .sui:
             let (referenceGasPrice, allCoins) = try await sui.getGasInfo(coin: coin)
             return .Sui(referenceGasPrice: referenceGasPrice, coins: allCoins)
-            
+
         case .polkadot:
             let gasInfo = try await dot.getGasInfo(fromAddress: coin.address)
             return .Polkadot(recentBlockHash: gasInfo.recentBlockHash, nonce: UInt64(gasInfo.nonce), currentBlockNumber: gasInfo.currentBlockNumber, specVersion: gasInfo.specVersion, transactionVersion: gasInfo.transactionVersion, genesisHash: gasInfo.genesisHash)
-            
+
         case .ethereum, .avalanche, .bscChain, .arbitrum, .base, .optimism, .polygon, .blast, .cronosChain:
             let service = try EvmServiceFactory.getService(forChain: coin.chain)
-            let (gasPrice, priorityFee, nonce) = try await service.getGasInfo(fromAddress: coin.address)
+            let baseFee = try await service.getBaseFee()
+            let (_, defaultPriorityFee, nonce) = try await service.getGasInfo(fromAddress: coin.address, mode: feeMode)
             let gasLimit = gasLimit ?? normalizeGasLimit(coin: coin, action: action)
-            let normalizedGasPrice = normalize(gasPrice, action: action)
-            return .Ethereum(maxFeePerGasWei: normalizedGasPrice, priorityFeeWei: normalizePriorityFee(priorityFee,coin.chain), nonce: nonce, gasLimit: gasLimit)
-            
+            let priorityFeesMap = try await service.fetchMaxPriorityFeesPerGas()
+            let priorityFee = priorityFeesMap[feeMode] ?? defaultPriorityFee
+            return .Ethereum(baseFee: baseFee, priorityFeeWei: priorityFee, nonce: nonce, gasLimit: gasLimit)
```

Comment:
if you do `return .Ethereum(baseFee:  (baseFee+priorityFee) , priorityFeeWei: priorityFee, nonce: nonce, gasLimit: gasLimit)` then we probably don't need to change it in `evm.swift`

Reply:
should work